lib/std/crypto/aes_cmac.zig

```diff
@@ -0,0 +1,171 @@
+/// RFC 4493 - The AES-CMAC Algorithm
+/// https://www.rfc-editor.org/rfc/rfc4493
+const std = @import("std");
+const Aes128 = std.crypto.core.aes.Aes128;
+const print = std.debug.print;
+
+pub fn AesCmac() type {
+    return struct {
+        const Self = @This();
+        var k1: [16]u8 = undefined;
+        var k2: [16]u8 = undefined;
+        var ctx: std.crypto.core.aes.AesEncryptCtx(Aes128) = undefined;
+
+        /// Create a new cmac context with the given key.
+        pub fn init(key: [16]u8) Self {
+            ctx = Aes128.initEnc(key);
+            k1 = [_]u8{0} ** 16;
+            k2 = [_]u8{0} ** 16;
+            ctx.encrypt(&k1, &k1);
+            makeKn(&k1, &k1);
+            makeKn(&k2, &k1);
+            return Self{};
+        }
+
+        /// Generates a cmac from the given message.
+        pub fn generate(self: Self, msg: []const u8) [16]u8 {
+            _ = self;
+
+            var flag: bool = undefined;
+            var Ml: [16]u8 = undefined;
+
+            // rounds
+            var n = (msg.len + 15) / 16;
+            if (n == 0) {
+                n = 1;
+                flag = false;
+            } else {
+                flag = if (msg.len % 16 == 0) true else false;
+            }
+
+            if (flag) {
+                Ml = xor(msg[16 * n - 16 ..], &k1);
+            } else {
+                var pad = padding(msg[16 * n - 16 ..], msg.len % 16);
+                Ml = xor(&pad, &k2);
+            }
+
+            var X: [16]u8 = [_]u8{0} ** 16;
+            var Y: [16]u8 = [_]u8{0} ** 16;
+            var i: usize = 0;
+            while (i < n - 1) : (i += 1) {
+                Y = xor(msg[16 * i ..], &X);
+                ctx.encrypt(&X, &Y);
+            }
+            Y = xor(&X, &Ml);
+
+            var t: [16]u8 = undefined;
+            ctx.encrypt(&t, &Y);
+
+            return t;
+        }
+
+        fn padding(lb: []const u8, len: usize) [16]u8 {
+            var pad: [16]u8 = [_]u8{0} ** 16;
+            var i: usize = 0;
+            while (i < 16) : (i += 1) {
+                if (i < len) {
+                    pad[i] = lb[i];
+                } else if (i == len) {
+                    pad[i] = 0x80;
+                } else {
+                    break;
+                }
+            }
+            return pad;
+        }
+
+        fn xor(a: []const u8, b: []const u8) [16]u8 {
+            var res: [16]u8 = [_]u8{0} ** 16;
+            for (b) |v, i| {
+                res[i] = a[i] ^ v;
+            }
+            return res;
+        }
+
+        // make temporary keys K1 and K2
+        fn makeKn(dst: []u8, src: []u8) void {
```

Comment:
Shouldn't `src`  be `const`? Maybe `dst` can be returned instead of `void`?

Reply:
> be represented by `Block.block_length` instead of hard-coded magic numbers.

Sorry, that these are not 'magic numbers' but constants that cannot 'accidentally' occur in the wrong length. AES-128 has 16 byte long blocks, so it is not helpful to define them by something else than constants. This has 2 advantages: 1. every auditor can immediately recognize what is meant and 2. nobody can accidentally work with wrong block lengths. By the way, this is a popular point of attack in cryptography.

Reply:
Please use `block_length` instead, so that it works with any block cipher, as CMAC is supposed to do. This is a constant (ex: `Aes128.Block.block_length` for AES-128).

Reply:
ok